### PR TITLE
feat(INJI-474): show popup for invalid QR

### PR DESCRIPTION
**Pull request by adityankannan-tw**
* at times a scanned QR code is neither online nor offline

### CODE REVIEW

**Comment by vharsh** [comment]
we'll remove some `console.log` and get back to this and re-add the `*typegen.ts` files.

File: machines/bleShare/scan/scanMachine.ts [hunk 1]
```diff
@@ -723,6 +732,9 @@ export const scanMachine =
     },
     {
       actions: {
+        offlineaccepted: () => console.log('>> offline QR accepted'),
```
Comment (vijay151096) [inline machines/bleShare/scan/scanMachine.ts:735]: do we need to log the QR acceptance or rejection , which would be seen in the UI itself, and moreover this log is more generic.

File: screens/Scan/ScanLayoutController.ts [hunk 2]
```diff
@@ -209,6 +209,7 @@ export function useScanLayout() {
   const isQrLoginDone = useSelector(scanService, selectIsQrLoginDone);
 
   useEffect(() => {
+    console.log('>> reviewing: ', isReviewing);
```
Comment (vijay151096) [inline screens/Scan/ScanLayoutController.ts:212]: can we remove the log here.

File: screens/Scan/ScanScreen.tsx [hunk 3]
```diff
@@ -151,13 +154,28 @@ export const ScanScreen: React.FC = () => {
...
           translationPath={'ScanScreen'}
           error="errors.storageLimitReached"
-          onDismiss={() => navigation.navigate(BOTTOM_TAB_ROUTES.home)}
+          onBackdropPress={() => navigation.navigate(BOTTOM_TAB_ROUTES.home)}
+        />
+      )
+    );
+  }
+
+  function invalidQR(): React.ReactNode {
```
Comment (vijay151096) [inline screens/Scan/ScanScreen.tsx:169]: functional component name usually starts with Capital Letter. can we refactor this as well.

File: screens/Scan/ScanScreen.tsx [hunk 4]
```diff
@@ -151,13 +154,28 @@ export const ScanScreen: React.FC = () => {
...
+    return (
+      !controller.isEmpty && (
+        <MessageOverlay
+          isVisible={controller.selectIsInvalid}
+          title={t('common.errors.genericError')}
+          // TODO: maybe add a message QR invalid
+          onButtonPress={() => navigation.navigate(BOTTOM_TAB_ROUTES.scan)}
+          buttonText={t('common:tryAgain')}
```
Comment (vijay151096) [inline screens/Scan/ScanScreen.tsx:177]: I think literals are tailed with . not :
this **buttonText={t('common:tryAgain')}** should be like **buttonText={t('common.tryAgain')}**

**Comment by vharsh** [comment]
Screenshot:

image

File: screens/Scan/ScanScreen.tsx (hunk 4 above)
Comment (vharsh) [reply screens/Scan/ScanScreen.tsx]: based on our conversations, this seems to be okay, let me know if further action is required on this

**Comment by vharsh** [comment]
[Update] I've changed the `Ok` button to `Cancel` as suggested in the ticket.

File: machines/bleShare/scan/scanMachine.ts [hunk 5]
```diff
@@ -1177,17 +1173,16 @@ export const scanMachine =
...
         isQrLogin: (_context, event) => {
-          let linkCode = '';
           try {
-            linkCode = event.params.substring(
-              event.params.indexOf('linkCode=') + 9,
-              event.params.indexOf('&'),
-            );
-            return linkCode !== null;
+            let linkCode = new URL(event.params);
+            // sample: 'inji://landing-page-name?linkCode=sTjp0XVH3t3dGCU&linkExpireDateTime=2023-11-09T06:56:18.482Z'
```
Comment (vijay151096) [inline machines/bleShare/scan/scanMachine.ts:1182]: sample is mentioned in line 1176 as well, we can remove this. and log message on catch can be like "Exception Occurred while ****** "

Comment (vharsh) [reply machines/bleShare/scan/scanMachine.ts:1182]: @vijay151096 That's a sample for the OpenIDqrCode i.e. offline QR code.

**Review by vijay151096 — APPROVED** [review]
looks good to me.

File: screens/Scan/ScanScreen.tsx [hunk 6]
```diff
@@ -1,11 +1,8 @@
 import React, {useEffect, useState} from 'react';
 import {useTranslation} from 'react-i18next';
-import {
-  ErrorMessageOverlay,
-  MessageOverlay,
-} from '../../components/MessageOverlay';
+import {MessageOverlay} from '../../components/MessageOverlay';
 import {QrScanner} from '../../components/QrScanner';
-import {Button, Centered, Column, Text} from '../../components/ui';
+import {Button, Centered, Column, Row, Text} from '../../components/ui';
```
Comment (Alka1703) [inline screens/Scan/ScanScreen.tsx:5]: Row is not getting used.  We can remove it.

File: screens/Scan/ScanScreen.tsx [hunk 7]
```diff
@@ -151,18 +151,44 @@ export const ScanScreen: React.FC = () => {
...
           translationPath={'ScanScreen'}
           error="errors.storageLimitReached"
-          onDismiss={() => navigation.navigate(BOTTOM_TAB_ROUTES.home)}
+          onBackdropPress={() => navigation.navigate(BOTTOM_TAB_ROUTES.home)}
         />
       )
     );
   }
 
+  function displayInvalidQR(): React.ReactNode {
```
Comment (KiruthikaJeyashankar) [inline screens/Scan/ScanScreen.tsx:166]: Can we rename this method name? This looks like we are displaying invalid QR 
cc: @sreenadh

Comment (vharsh) [reply screens/Scan/ScanScreen.tsx:166]: changed it to `displayInvalidQRpopup`

**Review by KiruthikaJeyashankar — APPROVED** [review]

**Comment by vharsh** [comment]
**Screenshots**:

**Comment by vharsh** [comment]
@vijay151096 This PR can be merged now --> https://github.com/mosip/inji/pull/999

**Review by Alka1703 — APPROVED** [review]

**Review by swatigoel — APPROVED** [review]

**Pull request merged**